### DISCUSSION TITLE
Support older GTK versions like 3.18.9

### DISCUSSION THREAD

**Discussion by caoliver** [dbeaver/dbeaver#21670] [Ideas]
### Description

In several panes I'm getting 'undefined symbol: gtk_menu_popup_at_pointer' in libswt-pi3-gtk-4960r6.so on right button click.  I'm getting 'undefined symbol: gtk_menu_popup_at_rect' on clicking on any of the arrow menu boxes.  The installed gtk-3 is version 3.18.9 if that matters.  Running readelf on libgtk-3.so show this symbol doesn't exist for this library.  Those symbols seem to exist in 3.24.31 though.  (checked on a container with different libraries)  At this point, an OS upgrade is a very painful ordeal as I have a number of stable boxes I really want to avoid changing as that breaks other things.

### DBeaver Version

Community Edition

### Operating System

Linux/Slackware 14.2

### Database and driver

This seems independent of the database.  In fact, I just killed off the saved configuration, so there are no remembered connections, and I see the same symptoms.

### Steps to reproduce

1. Start DBeaver.
2. Try and action that produces a popup menu.
3. Thud!

**Comment by LonwoLonwo** [comment]
Could you please provide the log files? You can find instructions here: https://dbeaver.com/docs/wiki/Log-files/. Thank you.

**Comment by caoliver** [comment]
On Tue, 17 Oct 2023 02:06:53 -0700
Anastasiya ***@***.***> wrote:

> Could you please provide the log files? You can find instructions here: https://dbeaver.com/docs/wiki/Log-files/. Thank you.

I'll indeed do so, but this seems to revolve around a breaking change in GTK-3's popup menu API.  I.e. the facility
provided in version 3.18.9 was deprecated and replaced with the calls given in my previous complaint.  I don't know
if this change which seems to be referenced in the SWT library can be papered over or not, and even so would this
just make the underlying code an unmaitainable rats nest?  At the moment, I'm contemplating a move to Slackware 15
which might well fix things, but it will take months to rebuild and test much of the local software to get things working
smoothly; hence my objection.  Not to belabor the point, but it would be nice if the FreeDesktop people put more
care and thought into backward compatibility.

Thanks so much for your consideration,

-- 
Christopher Oliver ***@***.***>

**Comment by caoliver** [comment]
On Tue, 17 Oct 2023 02:06:53 -0700
Anastasiya ***@***.***> wrote:

> Could you please provide the log files? You can find instructions here: https://dbeaver.com/docs/wiki/Log-files/. Thank you.

This is an excerpt from the log:

***WARNING: GTK+ version too old (micro mismatch)
***WARNING: SWT requires GTK 3.22.0
***WARNING: Detected: 3.18.9

This breaking change hurts existing installations.  Again, the GTK developers could have been more thoughtful.

-- 
Christopher Oliver ***@***.***>

**Comment by caoliver** [comment]
As an experiment, I went back in github history, and tried the 22.0.0 community
edition, and I don't see the undefined symbol issues.  I suppose I could "bisect"
through the release history and find out where GTK3 3.18.9 incompatibility started.

-- 
Christopher Oliver ***@***.***>

**Comment by caoliver** [comment]
Last version that works with GTK 3.18.9 is DBeaver 22.1.1.

-- 
Christopher Oliver ***@***.***>

**Reply by chirontt** [reply]
That DBeaver 22.1.1 embeds an ancient Eclipse 2022-03 version including the ancient SWT library that supports GTK 3.18.9. If this 22.1.1 version works on your system, why not continue to use it?

Alternatively, I've tried compiling the latest DBeaver code with that ancient Eclipse 2022-03 version, without success (of course, as expected.) Then similarly I've tried the latest release branch (release_23_2_3) without success. Ditto with the 23.2.2 release branch. And I give it up at this point.

**Comment by E1izabeth** [comment]
Thank you for investigation

**Comment by hrimthrusa** [comment]
Hi, please upgrade to GTK 3.22.0

**Comment by caoliver** [comment]
On Thu, 19 Oct 2023 06:20:27 -0700
Pavel Griva ***@***.***> wrote:

> Hi, please upgrade to GTK 3.22.0

If that were so easy!  I'm contemplating an upgrade of Slackware on my
workstations and servers, but this will mean months of aggravation and
breakage in systems that have been stable for years with just updates to
maintain security.  The GTK suite is tightly integrated, and often it is
impossible to change one component in isolation.  I am not even sure the
tool suite would build with the gcc toolchain on board assuming I go that
route.

-- 
Christopher Oliver ***@***.***>

**Reply by Destrolaric** [reply]
@caoliver Unfortunately, dependency on the newer version comes from the SWT graphical framework, which quite possibly started using some of the newer GTK functionality that is only compatible with a more recent version. We can't change the SWT version due to multiple Eclipse components on which we depend on using the current version of SWT. So the requirement for the GTK version is, unfortunately, here to stay.
